dbms/src/Operators/LocalSortTransformOp.cpp

```diff
@@ -0,0 +1,113 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <DataStreams/MergeSortingBlocksBlockInputStream.h>
+#include <DataStreams/SortHelper.h>
+#include <Interpreters/sortBlock.h>
+#include <Operators/LocalSortTransformOp.h>
+
+namespace DB
+{
+namespace
+{
+void cut(Block & block, size_t limit)
+{
+    if unlikely (!block)
+        return;
+    if (block.rows() <= limit)
+        return;
+    size_t pop_back_cnt = block.rows() - limit;
+    for (auto & col : block)
+    {
+        auto mutate_col = (*std::move(col.column)).mutate();
+        mutate_col->popBack(pop_back_cnt);
+        col.column = std::move(mutate_col);
+    }
+}
+} // namespace
+
+void LocalSortTransformOp::operatePrefix()
+{
+    header_without_constants = getHeader();
+    SortHelper::removeConstantsFromBlock(header_without_constants);
+    SortHelper::removeConstantsFromSortDescription(header, order_desc);
+}
+
+Block LocalSortTransformOp::getMergeOutput()
+{
+    assert(merge_impl);
+    Block block = merge_impl->read();
+    if likely (block)
+        SortHelper::enrichBlockWithConstants(block, header);
+    return block;
+}
+
+OperatorStatus LocalSortTransformOp::transformImpl(Block & block)
+{
+    switch (status)
+    {
+    case LocalSortStatus::PARTIAL:
+    {
+        /// If there were only const columns in sort description, then there is no need to sort, just cut block.
+        if (order_desc.empty())
+        {
+            cut(block, limit);
```

Comment:
Does this mean it will return all the input blocks after cut?

Reply:
yes, but here only a single block will be cutted, as with partial sort stream.
https://github.com/pingcap/tiflash/blob/e2f16fa997fbb551f5c709c1b455ff7e58ce5a03/dbms/src/DataStreams/MergeSortingBlockInputStream.cpp#L114-L117

Reply:
add `executeLocalSort` to append `LocalSortTransformOp` or `LocalLimitTransformOp`(order by const).